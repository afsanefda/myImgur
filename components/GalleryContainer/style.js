import styled from 'styled-components';

export const MainContainer = styled.section`
  height: 100vh;
`;

export const Title = styled.h1`
  font-size: 24px;
  text-align: center;
  color: #A65215;
`;

export const DropdownFilterContainer = styled.div`
  display: flex;
  justify-content: space-around;
`;

export const DropdownFilterItem = styled.div`
  display: flex;
  flex-direction: column;
  }
`;

export const FilterViral = styled.div`
  display: flex;
  align-items: center;
  }
`;

export const LoadingContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  }
`;

export const LoadingGif = styled.img`
  width: 200px;
  }
`;

export const LoadingTitle = styled.p`
  color: #A65215;
  }
`;

export const FilterViralBtn = styled.button`
  background: #F58535;
  color: #fff;
  box-shadow: none;
  border: none;
  padding: 5px 10px;
  border-radius: 6px;
  cursor: pointer;
  width: 150px;
  &:focus {
    outline: none;
  }
`;
