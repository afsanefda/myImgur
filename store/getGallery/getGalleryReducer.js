import {
  GET_GALLERY_FAILURE,
  GET_GALLERY_REQUEST,
  GET_GALLERY_SUCCESS,
} from '../constants';

const initialState = {
  loading: false,
  data: null,
  error: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    // Delete Application Request Cases
    case GET_GALLERY_REQUEST:
      return {
        ...state,
        loading: true,
      };
    case GET_GALLERY_SUCCESS:
      return {
        ...state,
        loading: false,
        data: action.payload,
      };
    case GET_GALLERY_FAILURE:
      return {
        ...state,
        loading: false,
        error: true,
      };

    default:
      return state;
  }
};
