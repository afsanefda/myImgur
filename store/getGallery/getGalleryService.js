import { baseURL } from '../request';

export const getGalleryServices = (page) => `${baseURL}/gallery/hot/viral/week/${page}?showViral=true&mature=true&album_previews=true`;
