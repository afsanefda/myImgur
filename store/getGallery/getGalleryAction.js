import {
  GET_GALLERY_FAILURE,
  GET_GALLERY_REQUEST,
  GET_GALLERY_SUCCESS,
} from '../constants';
// eslint-disable-next-line import/named
import { request } from '../request';
import { getGalleryServices } from './getGalleryService';

export const getGalleryAction = (page) => async (dispatch) => {
  dispatch({ type: GET_GALLERY_REQUEST });

  const response = await request.get(getGalleryServices(page));

  if (response.ok) {
    dispatch({ type: GET_GALLERY_SUCCESS, payload: response.data.data });
    return response;
  }

  dispatch({ type: GET_GALLERY_FAILURE });
  return false;
};
