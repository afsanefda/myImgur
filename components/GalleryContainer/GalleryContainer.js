// @flow
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getGalleryAction } from '../../store/getGallery/getGalleryAction';
import GalleryLists from '../GalleryList/GalleryList';
import {
  Title,
  DropdownFilterContainer,
  FilterViralBtn,
  FilterViral,
  DropdownFilterItem,
  MainContainer,
  LoadingContainer,
  LoadingGif,
  LoadingTitle,
} from './style';
import Dropdown from '../Dropdown/Dropdown';
import { sectionList, dateRangeList } from '../../webConfig';

const GalleryContainer = () => {
  const [viral, setViral] = useState(false);
  const [section, setSection] = useState('hot');
  const [dateRange, setDateRange] = useState('day');
  const [filterLoading, setFilterLoading] = useState(false);

  const loading = useSelector((state) => state.galleryList.loading);
  const error = useSelector((state) => state.galleryList.error);

  const dispatch = useDispatch();

  useEffect(() => {
    (async () => {
      setFilterLoading(true);
      await dispatch(getGalleryAction(true, 1, section, viral, dateRange));
      setFilterLoading(false);
    })();
  }, [viral,section,dateRange]); // eslint-disable-line

  const handleChangeViral = () => {
    setViral((prevState) => !prevState);
  };

  const handleChangeSection = (value) => {
    setSection(value);
  };

  const handleChangeWindow = (value) => {
    setDateRange(value);
  };

  const renderComponent = () => {
    if (error) {
      return <div>error</div>;
    } if (loading || filterLoading) {
      return (
        <LoadingContainer>
          <LoadingTitle>loading Data ...</LoadingTitle>
          <LoadingGif alt="MyImgur" src="/loading/loading.gif" />
        </LoadingContainer>
      );
    }
    return <GalleryLists viral={viral} section={section} dateRange={dateRange} />;
  };

  return (
    <MainContainer>
      <Title>
        Welcome to myImgur
      </Title>
      <DropdownFilterContainer>
        <DropdownFilterItem>
          <Dropdown
            defaultValue={section}
            list={sectionList}
            handleChangeFilter={handleChangeSection}
          />
          {
            section === 'top' && <Dropdown defaultValue={dateRange} list={dateRangeList} handleChangeFilter={handleChangeWindow} />
          }
        </DropdownFilterItem>
        <FilterViral>
          <FilterViralBtn type="button" onClick={handleChangeViral}>{viral ? 'hide viral images' : 'Show viral images' }</FilterViralBtn>
        </FilterViral>
      </DropdownFilterContainer>
      {renderComponent()}
    </MainContainer>
  );
};

export default GalleryContainer;
