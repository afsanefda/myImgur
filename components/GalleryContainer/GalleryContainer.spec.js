import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import createStore from 'redux-mock-store';
import waitForExpect from 'wait-for-expect';
import { getGalleryAction } from '../../store/getGallery/getGalleryAction';
import GalleryContainer from './GalleryContainer';

describe('<GalleryContainer />', () => {
  let rootElement;
  let testRenderer;
  let store;

  beforeEach(async () => {
    store = createStore([])({
      galleryList: {
        loading: false,
        data: [],
        error: false,
      },
    });
    store.dispatch = jest.fn();

    rootElement = () => (
      <Provider store={store}>
        <GalleryContainer />
      </Provider>
    );

    testRenderer = renderer.create(rootElement());
  });

  it('Should dispatch the getGalleryAction action on useEffect', () => {
    act(async () => {
      await waitForExpect(() => {
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(getGalleryAction(true, 1, 'hot', false, 'day'));
      });
    });
  });

  it('Should check viral button existence and on click action', () => {
    expect(testRenderer.root.findAllByProps({ 'data-test-id': 'viral-btn' }).length).toBe(1);
    expect(testRenderer.root.findAllByProps({ 'data-test-id': 'viral-btn' })[0].props.children).toBe('Show viral images');
    act(() => {
      testRenderer.root.findAllByProps({ 'data-test-id': 'viral-btn' })[0].props.onClick();
    });
    expect(testRenderer.root.findAllByProps({ 'data-test-id': 'viral-btn' })[0].props.children).toBe('hide viral images');
  });

  it('Should render correctly', () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
