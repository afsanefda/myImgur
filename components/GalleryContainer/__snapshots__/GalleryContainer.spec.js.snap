// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GalleryContainer /> Should render correctly 1`] = `
<section
  className="style__MainContainer-sc-15myhzc-0 eiHNiJ"
>
  <h1
    className="style__Title-sc-15myhzc-1 eZVBtB"
  >
    Welcome to myImgur
  </h1>
  <div
    className="style__DropdownFilterContainer-sc-15myhzc-2 iSKiKn"
  >
    <div
      className="style__DropdownFilterItem-sc-15myhzc-3 lgmho"
    >
      <div
        className="style__DropdownContainer-sc-1irrfnv-0 koxyMh"
      >
        <select
          className="style__DropdownSelect-sc-1irrfnv-1 cIWSah"
          onChange={[Function]}
          value="hot"
        >
          <option
            value="hot"
          >
            hot
          </option>
          <option
            value="top"
          >
            top
          </option>
          <option
            value="user"
          >
            user
          </option>
        </select>
      </div>
    </div>
    <div
      className="style__FilterViral-sc-15myhzc-4 jRgQlB"
    >
      <button
        className="style__FilterViralBtn-sc-15myhzc-8 foYqGx"
        data-test-id="viral-btn"
        onClick={[Function]}
        type="button"
      >
        Show viral images
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "height": "100vh",
        "outline": "none",
        "overflowY": "auto",
        "position": "relative",
      }
    }
    tabIndex={0}
  >
    <div
      style={
        Object {
          "height": "100%",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <div
        className="style__ListContainer-sc-1q4dzrq-1 iDuWjS virtuoso-grid-list"
        style={
          Object {
            "paddingBottom": "0px",
            "paddingTop": "0px",
          }
        }
      >
        <div
          className="style__ItemContainer-sc-1q4dzrq-0 hkdzmM virtuoso-grid-item"
        >
          <div
            className="style__ItemWrapper-sc-1nmiqb7-3 jXOryF"
          >
            <a
              href="/image/undefined"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <img
                className="style__ItemWrapperImg-sc-1nmiqb7-0 fVhjXU"
              />
              <p
                className="style__ItemWrapperTitle-sc-1nmiqb7-2 jaKrQv"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
